clang/tools/driver/cc1_main.cpp

```diff
@@ -469,6 +537,14 @@ Optional<int> CompileJobCache::tryReplayCachedResult(CompilerInstance &Clang) {
         OutputFile, &DiagOpts, /*MergeChildRecords*/ false, std::move(*OS));
     Diags.setClient(new ChainedDiagnosticConsumer(
         Diags.takeClient(), std::move(SerializedConsumer)));
+  } else {
+    // We always generate the serialized diagnostics so the key is independent
+    // of the presence of '--serialize-diagnostics'.
```

Comment:
Do we need to always generate the .d file when caching for the same reason?

Reply:
It's not clear to me how valuable that is, the implicit clang driver caching method is generating the .d file during dep-scanning (not cached), and when using the libclang API, generation of .d files should become unnecessary at some point.

Reply:
I'm fine with not optimizing for this case, but we should make it part of the cache key then, not just silently don't generate the .d file when replaying.  I suggest if there is a path, map it to "-", but if there is no path leave it empty.

Reply:
`.d` file is generated by dependency scanner so it should never be emitted from main compilation in cached build.

Reply:
This is true for in the driver+daemon, but not when using libclang APIs.

Reply:
> I suggest if there is a path, map it to "-", but if there is no path leave it empty.

Good point, I've updated the PR with such a change.